### ISSUE TITLE
tkinter PhotoImage data accepts bytes not only str

### ISSUE THREAD

**Issue by melassa** [python/typeshed#5127]
tkinter `PhotoImage` `data` accepts `bytes` (only?) under python 3.

`tk.PhotoImage(data=b'')`

`Mypy: Argument "data" to "PhotoImage" has incompatible type "bytes"; expected "str"`

**Comment by Akuli** [comment]
Most strings in tkinter can also be bytes. I'm not sure about whether we should allow bytes everywhere though; to me, it seems harmless but unnecessary for strings that are meant to be displayed in the UI.

**Comment by melassa** [comment]
"PhotoImage accepts a string of base64 encoded data"
Not only, it accepts also png and ppm bytes without base64 encoding:

**Comment by melassa** [comment]
Example of ppm bytes:

`    tk.PhotoImage(data=b'P6\n1 1\n255\n\x01\x02\x03')`

**Comment by Dnyarri** [comment]
> Example of ppm bytes:
> 
> ` tk.PhotoImage(data=b'P6\n1 1\n255\n\x01\x02\x03')`

Is this example supposed to produce anything but black (I see black with anything I try to feed as data PPM)?

Meanwhile, am I right supposing Tkinter does not support ASCII PPM (P3 type)? I have no info on this and no idea where to find it.

**Comment by Dnyarri** [comment]
> Tkinter supports png and gif images. To see the content of an image file as a byte string, use e.g. `print(open("image.png", "rb").read())`.

Thank you, I have no problems using PNG files. I'm just trying to reduce disk activity :-)

It would take a long to explain; briefly, I want some image filters GUI shell, that is, something that loads image, applies filter created by me, shows and saves result. You may see example on the bottom of this page
[ example on the bottom of this page](https://dnyarri.github.io/povthread.html). And I want things to be as pure-Python as possible. So far I do all these preview things by saving tempfile PNG via PyPNG, then viewing tempfile via Tkinter. Everything works but I want to remove this intermediate tempfile, for which I wanted to use PPM as "data". Well, I'm just a novice, so I'm slow, sorry, but finally I seem to find the solution (at least I just loaded PNG, converting it to list for my needs then to bytearray and finally byte for tkinter needs, and it is displayed ok. So I guess I just need to rewrite all my filter template, and I get some almost visual filter builder, like old Adobe Photoshop Filter Factory, only without Adobe Photoshop part and much more capable).

Thanks for help!

**Comment by Dnyarri** [comment]
> You are right: ASCII PPM does not work. Only binary.

Thank you for clarification. I found one line regarding "prefering" binaries, but it was about ten years old and not too specific. Well, now I know this for sure, thank you! Looks like I finally managed to produce correct bin PPM "data" for tkinter (well, I don't care about line length so I put all the pixels in one line to see what happens. Happily, tkinter doesn't seem to care about it either).
